### PR TITLE
registry auth support for cluster up

### PR DESCRIPTION
**Pull request by bparees**
fixes bug https://bugzilla.redhat.com/show_bug.cgi?id=1592936

fixes https://github.com/openshift/origin/issues/20022

### CODE REVIEW

File: pkg/image/apiserver/apiserver.go [hunk 1]
```diff
@@ -119,11 +125,36 @@ func (c *completedConfig) V1RESTStorage() (map[string]rest.Storage, error) {
...
+	var err error
+	tlsConfig.RootCAs, err = x509.SystemCertPool()
+	if err != nil {
+		return nil, fmt.Errorf("unable to get system cert pool for default transport for image importing: %v", err)
+	}
+	if tlsConfig.RootCAs == nil {
+		tlsConfig.RootCAs = x509.NewCertPool()
+	}
+
+	if _, err := os.Stat("/etc/origin/master/imageimportcerts/cert.pem"); !os.IsNotExist(err) {
```
Comment (mfojtik) [inline pkg/image/apiserver/apiserver.go]: i think this will have to be plumbed from master config, right @deads2k ?

Comment (deads2k) [reply pkg/image/apiserver/apiserver.go]: > i think this will have to be plumbed from master config, right @deads2k ?

Yeah.  I suspect WIP really means it here and he didn't mean to cause us heart failure. :)

Comment (bparees) [reply pkg/image/apiserver/apiserver.go]: it really doesn't.  /etc/origin/master is always mounted already, this is taking advantage of that to say "if you want to provide additional inputs, drop them in this dir".

Comment (bparees) [reply pkg/image/apiserver/apiserver.go]: i've updated this to get a path from the master-config instead of hardcoding one.

There is also no default, so you have to set it explicitly in master-config and then ensure the file exists/is mounted into the apiserver pod.

**Comment by bparees** [comment]
@mfojtik @deads2k I think this is ready for review in terms of cluster up changes and the master-config changes.

@dmage @legionus imageimport transport setup changes ready for review.

File: pkg/cmd/server/apis/config/types.go [hunk 2]
```diff
@@ -526,6 +526,10 @@ type ImagePolicyConfig struct {
 	// is exposed externally. The value is used in 'publicDockerImageRepository'
 	// field in ImageStreams. The value must be in "hostname[:port]" format.
 	ExternalRegistryHostname string
+
+	// AdditionalTrustedCAs is a path to a pem file containing additional CAs that
+	// should be trusted during imagestream import.
+	AdditionalTrustedCAs string
```
Comment (deads2k) [inline pkg/cmd/server/apis/config/types.go]: CA singular.  Doc should indicate that is can be a bundle.

Doing it like this means that the trust domain of all CA bundles specified here is uniform.  Any can sign for any names related to image importing.  I'm ok with that.  Just calling it out.

Comment (deads2k) [inline pkg/cmd/server/apis/config/types.go]: Add to `ResolveMasterConfigPaths`

File: pkg/image/apiserver/apiserver.go [hunk 3]
```diff
@@ -116,11 +125,41 @@ func (c *completedConfig) V1RESTStorage() (map[string]rest.Storage, error) {
...
+	if err != nil {
+		return nil, fmt.Errorf("unable to get system cert pool for default transport for image importing: %v", err)
+	}
+	if tlsConfig.RootCAs == nil {
+		tlsConfig.RootCAs = x509.NewCertPool()
+	}
+
+	glog.V(2).Infof("Image import using additional CA path: %s", c.ExtraConfig.AdditionalTrustedCAs)
+	if len(c.ExtraConfig.AdditionalTrustedCAs) != 0 {
+		if _, err := os.Stat(c.ExtraConfig.AdditionalTrustedCAs); !os.IsNotExist(err) {
```
Comment (deads2k) [inline pkg/image/apiserver/apiserver.go]: perform this validation in the master config validation itself.

File: pkg/oc/clusteradd/componentinstall/create_secret.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+	secret  DockerConfigSecret
+	image   string
+	baseDir string
+}
+
+func (opt *installReadyDockerConfigSecret) Name() string {
+	return opt.secret.Name
+}
+
+func (opt *installReadyDockerConfigSecret) Install(dockerClient dockerhelper.Interface) error {
```
Comment (deads2k) [inline pkg/oc/clusteradd/componentinstall/create_secret.go]: need to be indempotent

File: pkg/oc/clusteradd/componentinstall/create_secret.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/openshift/origin/pkg/oc/clusterup/coreinstall/kubeapiserver"
+	"github.com/openshift/origin/pkg/oc/clusterup/docker/dockerhelper"
+	"github.com/openshift/origin/pkg/oc/clusterup/docker/run"
+	"github.com/openshift/origin/pkg/oc/errors"
+)
+
+type DockerConfigSecret struct {
```
Comment (deads2k) [inline pkg/oc/clusteradd/componentinstall/create_secret.go]: I don't see us getting super strict on what components can be added, but are you sure this one is worth the effort?  Seems fairly narrow and easy to add after the fact

File: pkg/oc/clusteradd/components/default-imagestreams/default_imagestreams.go [hunk 6]
```diff
@@ -20,6 +20,18 @@ func (c *RHELImageStreamsComponentOptions) Name() string {
 }
 
 func (c *RHELImageStreamsComponentOptions) Install(dockerClient dockerhelper.Interface) error {
+	secretComponent := componentinstall.DockerConfigSecret{
```
Comment (deads2k) [inline pkg/oc/clusteradd/components/default-imagestreams/default_imagestreams.go]: Oh ordering sadness.  Is this really an independent thing to install then?  Is it narrowly tailored for this?  if so, just include in this package.

File: pkg/oc/clusteradd/components/registry/registry_install.go [hunk 7]
```diff
@@ -33,24 +32,6 @@ func (r *RegistryComponentOptions) Name() string {
 	return "openshift-image-registry"
 }
 
-// ensureRemoteRegistryStoragePermissions ensures the remote host directory for registry storage have write access permissions
-// so the registry can successfully write data into it.
-// TODO: This is a remote docker snowflake
-func (r *RegistryComponentOptions) ensureRemoteRegistryStoragePermissions(dir string, dockerClient dockerhelper.Interface) error {
```
Comment (deads2k) [inline pkg/oc/clusteradd/components/registry/registry_install.go]: yay.  @mfojtik why was this here?

File: pkg/image/apiserver/apiserver.go (hunk 3 above)
Comment (bparees) [reply pkg/image/apiserver/apiserver.go]: masterconfig may be shared between pods, not all pods need this file, so doing the validation globally seems like a mistake?

File: pkg/oc/clusteradd/componentinstall/create_secret.go (hunk 5 above)
Comment (bparees) [reply pkg/oc/clusteradd/componentinstall/create_secret.go]: what do you propose instead?

**Comment by bparees** [comment]
@deads2k comments replied to or addressed in `david1` commit

File: pkg/cmd/server/apis/config/types.go (hunk 2 above)
Comment (bparees) [reply pkg/cmd/server/apis/config/types.go]: updated

Comment (bparees) [reply pkg/cmd/server/apis/config/types.go]: added

File: pkg/oc/clusteradd/componentinstall/create_secret.go (hunk 4 above)
Comment (bparees) [reply pkg/oc/clusteradd/componentinstall/create_secret.go]: done

File: pkg/oc/clusteradd/componentinstall/create_secret.go (hunk 5 above)
Comment (bparees) [reply pkg/oc/clusteradd/componentinstall/create_secret.go]: since i've moved this per your other comment, not sure if this comment is still applicable or not

File: pkg/oc/clusteradd/components/default-imagestreams/default_imagestreams.go (hunk 6 above)
Comment (bparees) [reply pkg/oc/clusteradd/components/default-imagestreams/default_imagestreams.go]: moved

**Comment by enj** [comment]
@openshift/sig-security certs!

**Comment by bparees** [comment]
@deads2k further thoughts on where filepath validation should be done and how fatal it should be?

**Comment by bparees** [comment]
@deads2k validation logic updated.

File: pkg/cmd/server/apis/config/v1/types.go [hunk 8]
```diff
@@ -411,6 +411,9 @@ type ImagePolicyConfig struct {
 	// is exposed externally. The value is used in 'publicDockerImageRepository'
 	// field in ImageStreams. The value must be in "hostname[:port]" format.
 	ExternalRegistryHostname string `json:"externalRegistryHostname,omitempty"`
+	// AdditionalTrustedCA is a path to a pem bundle file containing additional CAs that
+	// should be trusted during imagestream import.
+	AdditionalTrustedCA string `json:"AdditionalTrustedCA,omitempty"`
```
Comment (deads2k) [inline pkg/cmd/server/apis/config/v1/types.go]: lowercase leading letter.

File: pkg/cmd/server/apis/config/validation/master.go [hunk 9]
```diff
@@ -459,6 +460,11 @@ func ValidateImagePolicyConfig(config configapi.ImagePolicyConfig, fldPath *fiel
 			}
 		}
 	}
+	if len(config.AdditionalTrustedCA) != 0 {
```
Comment (deads2k) [inline pkg/cmd/server/apis/config/validation/master.go]: `ValidateFile`

File: pkg/cmd/server/origin/openshift_apiserver.go [hunk 10]
```diff
@@ -95,6 +95,7 @@ type OpenshiftAPIExtraConfig struct {
 	RegistryHostnameRetriever          imageapi.RegistryHostnameRetriever
 	AllowedRegistriesForImport         *configapi.AllowedRegistries
 	MaxImagesBulkImportedPerRepository int
+	AdditionalTrustedCA                string
```
Comment (deads2k) [inline pkg/cmd/server/origin/openshift_apiserver.go]: At this point we should have bytes

File: pkg/image/apiserver/apiserver.go [hunk 11]
```diff
@@ -43,6 +51,7 @@ type ExtraConfig struct {
 	RegistryHostnameRetriever          imageapi.RegistryHostnameRetriever
 	AllowedRegistriesForImport         *configapi.AllowedRegistries
 	MaxImagesBulkImportedPerRepository int
+	AdditionalTrustedCA                string
```
Comment (deads2k) [inline pkg/image/apiserver/apiserver.go]: bytes

File: pkg/cmd/server/origin/openshift_apiserver.go (hunk 10 above)
Comment (bparees) [reply pkg/cmd/server/origin/openshift_apiserver.go]: what's the motivation for that?

File: pkg/image/apiserver/apiserver.go (hunk 11 above)
Comment (deads2k) [reply pkg/image/apiserver/apiserver.go]: > bytes

Actually, the CA type itself.

File: pkg/image/apiserver/apiserver.go [hunk 12]
```diff
@@ -116,11 +125,41 @@ func (c *completedConfig) V1RESTStorage() (map[string]rest.Storage, error) {
...
+	}
+
+	glog.V(2).Infof("Image import using additional CA path: %s", c.ExtraConfig.AdditionalTrustedCA)
+	if len(c.ExtraConfig.AdditionalTrustedCA) != 0 {
+		if _, err := os.Stat(c.ExtraConfig.AdditionalTrustedCA); !os.IsNotExist(err) {
+			caData, err := ioutil.ReadFile(c.ExtraConfig.AdditionalTrustedCA)
+			if err != nil {
+				return nil, fmt.Errorf("failed to read ca bundle for image importing: %v", err)
+			}
+			tlsConfig.RootCAs.AppendCertsFromPEM(caData)
```
Comment (mrogers950) [inline pkg/image/apiserver/apiserver.go]: AppendCertsFromPEM() could return false.

File: pkg/cmd/server/origin/openshift_apiserver.go (hunk 10 above)
Comment (deads2k) [reply pkg/cmd/server/origin/openshift_apiserver.go]: > what's the motivation for that?

containability and testability.  We have a serializable config focus on human users, but the next step is to transform that into a runnable, nonserializable struct.  This is that struct.  Passing serializeable data that needs more transformation is against that goal.

File: pkg/image/apiserver/apiserver.go (hunk 11 above)
Comment (bparees) [reply pkg/image/apiserver/apiserver.go]: where can i find that type?

File: pkg/cmd/server/apis/config/v1/types.go (hunk 8 above)
Comment (bparees) [reply pkg/cmd/server/apis/config/v1/types.go]: addressed in `david2`

File: pkg/cmd/server/apis/config/validation/master.go (hunk 9 above)
Comment (bparees) [reply pkg/cmd/server/apis/config/validation/master.go]: addressed in `david2`

File: pkg/cmd/server/origin/openshift_apiserver.go (hunk 10 above)
Comment (bparees) [reply pkg/cmd/server/origin/openshift_apiserver.go]: addressed in `david2`

File: pkg/image/apiserver/apiserver.go (hunk 11 above)
Comment (bparees) [reply pkg/image/apiserver/apiserver.go]: addressed in `david2`

File: pkg/image/apiserver/apiserver.go (hunk 12 above)
Comment (bparees) [reply pkg/image/apiserver/apiserver.go]: addressed in `david2`

File: pkg/oc/clusteradd/components/default-imagestreams/create_secret.go [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+
+	dockerConfigJsonBytes, err := ioutil.ReadFile(path.Join(home, ".docker", "config.json"))
+	if err != nil {
+		return fmt.Errorf("Error reading $HOME/.docker/config.json: %v", err)
+	}
+
+	glog.Infof("Installing %q", opt.Name())
+
+	contentToCopy := map[string][]byte{
+		"kubeconfig.kubeconfig": clusterAdminConfigBytes,
```
[the remark this answers is not in the document]
Comment (bparees) [reply pkg/oc/clusteradd/components/default-imagestreams/create_secret.go]: addressed in `david2` but fyi this was copied from apply_list.go so that has the same issue (as does apply_template)

**Comment by deads2k** [comment]
/lgtm

please squash in the fixup commits.

**Comment by bparees** [comment]
[(5) Tolerate auth-enabled redhat registry [builds]](https://trello.com/c/Howwzs2c/1492-5-tolerate-auth-enabled-redhat-registry-builds)
